ruler/scheduler.go

```diff
@@ -0,0 +1,229 @@
+	prometheus.MustRegister(queueLength)
+	prometheus.MustRegister(blockedWorkers)
+	prometheus.MustRegister(configsRequestDuration)
+	prometheus.MustRegister(totalConfigs)
+}
+
+type workItem struct {
+	userID    string
+	rules     []rules.Rule
+	scheduled time.Time
+}
+
+// Key implements Op
+func (w workItem) Key() string {
+	return w.userID
+}
+
+// Priority implements Op
+func (w workItem) Priority() int64 {
+	return -w.scheduled.Unix()
+}
+
+// Defer returns a copy of this work item, rescheduled to a later time.
+func (w workItem) Defer(interval time.Duration) workItem {
+	return workItem{w.userID, w.rules, w.scheduled.Add(interval)}
+}
+
+type scheduler struct {
+	configsAPI         configsAPI // XXX: Maybe make this an interface ConfigSource or similar.
+	evaluationInterval time.Duration
+	q                  *util.PriorityQueue
+
+	// All the configurations that we have. Only used for instrumentation.
+	cfgs map[string]cortexConfig
+
+	pollInterval time.Duration
+	lastPolled   time.Time
+	pollMutex    sync.RWMutex
+
+	quit chan struct{}
+	wait sync.WaitGroup
+}
+
+// newScheduler makes a new scheduler.
+func newScheduler(configsAPI configsAPI, evaluationInterval, pollInterval time.Duration) scheduler {
+	return scheduler{
+		configsAPI:         configsAPI,
+		evaluationInterval: evaluationInterval,
+		pollInterval:       pollInterval,
+		q:                  util.NewPriorityQueue(),
+		cfgs:               map[string]cortexConfig{},
+	}
+}
+
+// Run polls the source of configurations for changes.
+func (s *scheduler) Run() {
+	log.Debugf("Scheduler started")
+	s.wait.Add(1)
+	defer s.wait.Done()
+	// Load initial set of all configurations before polling for new ones.
+	s.addNewConfigs(time.Now(), s.loadAllConfigs())
+	ticker := time.NewTicker(s.pollInterval)
+	for {
+		select {
+		case now := <-ticker.C:
+			err := s.updateConfigs(now)
+			if err != nil {
+				log.Warnf("Error updating configs: %v", err)
+			}
+		case <-s.quit:
+			ticker.Stop()
+		}
+	}
+}
+
+func (s *scheduler) Stop() {
+	close(s.quit)
+	s.q.Close()
+	s.wait.Wait()
+	log.Debugf("Scheduler stopped")
+}
+
+// Load the full set of configurations from the server, retrying with backoff
+// until we can get them.
+func (s *scheduler) loadAllConfigs() map[string]cortexConfig {
+	backoff := minBackoff
+	dawnOfTime := time.Unix(0, 0)
+	for {
+		cfgs, err := s.poll(dawnOfTime)
+		if err == nil {
+			log.Debugf("Found %d configurations in initial load", len(cfgs))
+			return cfgs
+		}
+		log.Warnf("Error fetching all configurations, backing off: %v", err)
+		time.Sleep(backoff)
+		backoff *= 2
+		if backoff > maxBackoff {
+			backoff = maxBackoff
+		}
+	}
+}
+
+func (s *scheduler) updateConfigs(now time.Time) error {
+	cfgs, err := s.poll(now)
+	if err != nil {
+		return err
+	}
+	s.addNewConfigs(now, cfgs)
+	return nil
+}
+
+func (s *scheduler) poll(now time.Time) (map[string]cortexConfig, error) {
+	s.pollMutex.RLock()
+	interval := now.Add(-skewCorrection).Sub(s.lastPolled)
```

Comment:
Hmmm, wouldn't it be better if you got all configs since the last config, where that is defined as max timestamp of last config?  Then you wouldn't have to worry about skew, and you could use absolute times.